src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -505,18 +513,28 @@ static bool useHeavyMonitors() {
 // The interpreter and compiler assembly code tries to lock using the fast path
 // of this algorithm. Make sure to update that code if the following function is
 // changed. The implementation is extremely sensitive to race condition. Be careful.
-
 void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* current) {
+  assert(current == JavaThread::current(), "must lock on the current thread");
+  enter(obj, lock, current, current);
+}
+
+void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* locking_thread, JavaThread* current) {
```

Comment:
Do we actually need to pass in the current thread? What is it used for - ResourceMarks?

Reply:
It is definitely an alternative to simply use Thread::current. That is all it is used for throughout this code.

Reply:
Yes an alternative could be at the beginning of 'enter' do:
Thread* current = Thread::current();

You have a comment below about locking_thread not necessarily being current.